lib/at_cmd_parser/at_cmd_parser.c

```diff
@@ -0,0 +1,340 @@
+				/* Only first parameter in the string can be
+				   notification ID, (eg +CEREG:) */
+				set_new_state(NOTIFICATION);
+			} else if ((index > 0) &&
+				is_notification(*str)) {
+				/* If notifications is detected later in the
+				   string we should stop parsing and return
+				   EAGAIN */
+				break;
+			} else if (is_number(*str)) {
+				set_new_state(NUMBER);
+
+			} else if (is_dblquote(*str)) {
+				set_new_state(STRING);
+				str++;
+			} else if (is_array_start(*str)) {
+				set_new_state(ARRAY);
+				str++;
+			} else if (is_lfcr(*str) &&
+			          (prev_state == NUMBER)) {
+				/* If \n or \r is detected in the string and the
+				   previous param was a number we assume the
+				   next parameter is PDU data */
+
+				while (is_lfcr(*str)) {
+					str++;
+				}
+
+				set_new_state(SMS_PDU);
+			} else if ((index == 0) &&
+				    !is_notification(*str)) {
+				/* If the string start without an notification
+				   ID, we treat the whole string as one string
+				   parameter */
+				set_new_state(STRING);
+			} else {
+				if (is_separator(*str)) {
+				/* If a separator is detected we have detected
+				   and empty optional parameter */
+					set_new_state(OPTIONAL);
+				} else {
+				/* The rule set is exhausted, and cannot
+				   continue. Break the loop and return an error
+				*/
+					break;
+				}
+			}
+		} else if (state == NOTIFICATION) {
+			const char *start_ptr = str++;
+
+			while (is_valid_notification_char(*str)) {
+				str++;
+			}
+
+			at_params_string_put(list,
+					     index, start_ptr,
+					     str - start_ptr);
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == OPTIONAL) {
+			at_params_empty_put(list, index);
+			str++;
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == STRING) {
+			const char *start_ptr = str;
+
+			while (!is_dblquote(*str) &&
+			       !is_terminated(*str)) {
+				str++;
+			}
+
+			at_params_string_put(list,
+					     index,
+					     start_ptr, str - start_ptr);
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == ARRAY) {
+			size_t i = 0;
+			u32_t  tmparray[16];
+
+			tmparray[i++] = atoi(str++);
+
+			while (!is_array_stop(*str) &&
+			       !is_terminated(*str)) {
+
+				if (is_separator(*str)) {
+					tmparray[i++] = atoi(++str);
+				} else {
+					str++;
+				}
+			}
+
+			at_params_array_put(list, index,
+					    tmparray,
+					    i * sizeof(u32_t));
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == NUMBER) {
+			int value = atoi(str);
+
+			if (value <= USHRT_MAX) {
+				at_params_short_put(list, index, (u16_t)value);
+			} else {
+				at_params_int_put(list, index, value);
+			}
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == SMS_PDU) {
+			const char *start_ptr = str;
+
+			while (isxdigit(*str)) {
+				str++;
+			}
+
+			at_params_string_put(list,
+					     index,
+					     start_ptr, str - start_ptr);
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == SEARCH_NEXT_PARAM) {
+
```

Comment:
Stray empty line.

Reply:
Between the else if and the while? It was on purpose, but removed now.